File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/03．持续危险区/02．线段危险区.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";
import type { 技能提示圈配置 } from "../../02．通用函数/16．技能提示圈工厂";
import { 创建技能提示圈 } from "../../02．通用函数/16．技能提示圈工厂";

const jass = require("jass.common") as any;

const GetUnitX = jass.GetUnitX as (u: any) => number;
const GetUnitY = jass.GetUnitY as (u: any) => number;

const { CosBJ, SinBJ } = require("lib.扩展函数.BJ函数.12．数学函数") as {
  CosBJ: (this: void, degrees: number) => number;
  SinBJ: (this: void, degrees: number) => number;
};
const { addPeriodicCallback, removePeriodicCallback, getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void, variable?: any) => void, variable?: any) => number;
  removePeriodicCallback: (this: void, id: number) => void;
  getServerTime: (this: void) => number;
};
const { isValidUnit } = require("lib.扩展函数.自定义扩展函数.02．条件判断函数") as {
  isValidUnit: (this: void, unit: any) => boolean;
};

export interface 线段危险区参数 {
  清理?: 机制清理篮子;
  名称: string;
  起点X: number;
  起点Y: number;
  方向角: number;
  长度: number;
  宽度: number;
  持续秒: number;
  Tick间隔毫秒?: number;
  周期秒?: number;
  /** 传给单位列表与生命周期回调的业务上下文。 */
  变量?: any;
  /** 真正跨过线段中心线的最小触发间隔。 */
  穿越防抖秒?: number;
  单位列表: (this: void, 变量?: any) => any[];
  /** 仅配置提示表现；坐标始终强制使用线段危险区的真实起点，不能覆盖为模型中点。 */
  提示圈?: 技能提示圈配置 | false;
  on进入?: (this: void, 单位: any, 变量?: any) => void;
  on离开?: (this: void, 单位: any, 变量?: any) => void;
  on穿越?: (this: void, 单位: any, 变量?: any) => void;
  on周期?: (this: void, 单位: any, 变量?: any) => void;
  on结束?: (this: void) => void;
}

export interface 线段危险区实例 {
  停止(): void;
}

interface 单位线段状态 {
  在内部: boolean;
  下次周期Ms: number;
  已采样: boolean;
  上次前向距离: number;
  上次侧向符号: number;
  下次穿越Ms: number;
}

function 绝对值(this: void, value: number): number {
  return value >= 0 ? value : -value;
}

function 前向距离在范围内(this: void, value: number, 长度: number): boolean {
  return value >= 0 && value <= 长度;
}

class 线段危险区实现 implements 线段危险区实例 {
  private 参数: 线段危险区参数;
  private 到期Ms: number;
  private 单位状态表: Record<number, 单位线段状态 | undefined> = {};
  private 前向X: number;
  private 前向Y: number;
  private 右向X: number;
  private 右向Y: number;
  private Tick回调ID = 0;
  private 已停止 = false;

  constructor(参数: 线段危险区参数) {
    this.参数 = 参数;
    this.到期Ms = getServerTime() + 参数.持续秒 * 1000;
    this.前向X = CosBJ(参数.方向角);
    this.前向Y = SinBJ(参数.方向角);
    this.右向X = CosBJ(参数.方向角 - 90);
    this.右向Y = SinBJ(参数.方向角 - 90);
    this.创建提示圈();
    this.Tick回调ID = addPeriodicCallback(参数.Tick间隔毫秒 ?? 100, on线段危险区Tick, this);
  }

  推进(now: number): void {
    if (this.已停止) return;
    if (now >= this.到期Ms) {
      this.停止();
      return;
    }
    const 单位列表 = this.参数.单位列表(this.参数.变量);
    for (let i = 0; i < 单位列表.length; i++) {
      const 单位 = 单位列表[i];
      if (!isValidUnit(单位)) continue;
      this.推进单位(now, 单位);
    }
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    if (this.Tick回调ID !== 0) {
      removePeriodicCallback(this.Tick回调ID);
      this.Tick回调ID = 0;
    }
    if (this.参数.on结束 != null) this.参数.on结束();
  }

  private 推进单位(now: number, 单位: any): void {
    const id = jass.GetHandleId(单位);
    const dx = GetUnitX(单位) - this.参数.起点X;
    const dy = GetUnitY(单位) - this.参数.起点Y;
    const 前向距离 = dx * this.前向X + dy * this.前向Y;
    const 横向距离 = dx * this.右向X + dy * this.右向Y;
    const 当前在内部 = this.是否在内部(GetUnitX(单位), GetUnitY(单位));
    let 状态 = this.单位状态表[id];
    if (状态 == null) {
      状态 = {
        在内部: false,
        下次周期Ms: 0,
        已采样: false,
        上次前向距离: 前向距离,
        上次侧向符号: 0,
        下次穿越Ms: 0,
      };
      this.单位状态表[id] = 状态;
    }
    const 当前侧向符号 = 横向距离 < 0 ? -1 : (横向距离 > 0 ? 1 : 0);
    const 真正跨线 = 状态.已采样
      && 当前侧向符号 !== 0
      && 状态.上次侧向符号 !== 0
      && 当前侧向符号 !== 状态.上次侧向符号
      && 前向距离在范围内(状态.上次前向距离, this.参数.长度)
      && 前向距离在范围内(前向距离, this.参数.长度)
      && now >= 状态.下次穿越Ms;
    状态.已采样 = true;
    状态.上次前向距离 = 前向距离;
    if (当前侧向符号 !== 0) 状态.上次侧向符号 = 当前侧向符号;
    if (当前在内部 && !状态.在内部) {
      状态.在内部 = true;
      状态.下次周期Ms = now;
      if (this.参数.on进入 != null) this.参数.on进入(单位, this.参数.变量);
    } else if (!当前在内部 && 状态.在内部) {
      状态.在内部 = false;
      if (this.参数.on离开 != null) this.参数.on离开(单位, this.参数.变量);
    }
    if (真正跨线) {
      状态.下次穿越Ms = now + (this.参数.穿越防抖秒 ?? 0) * 1000;
      if (this.参数.on穿越 != null) this.参数.on穿越(单位, this.参数.变量);
    }
    if (当前在内部 && this.参数.on周期 != null && now >= 状态.下次周期Ms) {
      this.参数.on周期(单位, this.参数.变量);
      状态.下次周期Ms = now + (this.参数.周期秒 ?? 1) * 1000;
    }
  }

  private 是否在内部(x: number, y: number): boolean {
    const dx = x - this.参数.起点X;
    const dy = y - this.参数.起点Y;
    const 前向距离 = dx * this.前向X + dy * this.前向Y;
    if (前向距离 < 0 || 前向距离 > this.参数.长度) return false;
    const 横向距离 = dx * this.右向X + dy * this.右向Y;
    return 绝对值(横向距离) <= this.参数.宽度 * 0.5;
  }

  private 创建提示圈(): void {
    if (this.参数.提示圈 === false) return;
    创建技能提示圈({
      类型: "矩形",
      宽度: this.参数.宽度,
      长度: this.参数.长度,
      朝向: this.参数.方向角,
      持续时间: this.参数.持续秒,
      ...(this.参数.提示圈 ?? {}),
      X: this.参数.起点X,
      Y: this.参数.起点Y,
    });
  }
}

function on线段危险区Tick(this: void, variable?: any): void {
  const 实例 = variable as 线段危险区实现 | undefined;
  if (实例 != null) 实例.推进(getServerTime());
}

export function 创建线段危险区(this: void, 参数: 线段危险区参数): 线段危险区实例 {
  const 实例 = new 线段危险区实现(参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(参数.名称, function 线段危险区清理(this: void): void {
      实例.停止();
    });
  }
  return 实例;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/03．持续危险区/02．线段危险区.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____on_7EBF_6BB5_5371_9669_533ATick, getServerTime
local ____16_FF0E_6280_80FD_63D0_793A_5708_5DE5_5382 = require("系统.03．技能系统.00．技能模板+函数.02．通用函数.16．技能提示圈工厂")
local _____521B_5EFA_6280_80FD_63D0_793A_5708 = ____16_FF0E_6280_80FD_63D0_793A_5708_5DE5_5382["创建技能提示圈"]
function ____on_7EBF_6BB5_5371_9669_533ATick(variable)
    local _____5B9E_4F8B = variable
    if _____5B9E_4F8B ~= nil then
        _____5B9E_4F8B["推进"](
            _____5B9E_4F8B,
            getServerTime()
        )
    end
end
local jass = require("jass.common")
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local ____require_result_0 = require("lib.扩展函数.BJ函数.12．数学函数")
local CosBJ = ____require_result_0.CosBJ
local SinBJ = ____require_result_0.SinBJ
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_1.addPeriodicCallback
local removePeriodicCallback = ____require_result_1.removePeriodicCallback
getServerTime = ____require_result_1.getServerTime
local ____require_result_2 = require("lib.扩展函数.自定义扩展函数.02．条件判断函数")
local isValidUnit = ____require_result_2.isValidUnit
local function _____7EDD_5BF9_503C(value)
    return value >= 0 and value or -value
end
local function _____524D_5411_8DDD_79BB_5728_8303_56F4_5185(value, _____957F_5EA6)
    return value >= 0 and value <= _____957F_5EA6
end
local _____7EBF_6BB5_5371_9669_533A_5B9E_73B0 = __TS__Class()
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.name = "线段危险区实现"
function _____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype.____constructor(self, _____53C2_6570)
    self["单位状态表"] = {}
    self["Tick回调ID"] = 0
    self["已停止"] = false
    self["参数"] = _____53C2_6570
    self["到期Ms"] = getServerTime() + _____53C2_6570["持续秒"] * 1000
    self["前向X"] = CosBJ(_____53C2_6570["方向角"])
    self["前向Y"] = SinBJ(_____53C2_6570["方向角"])
    self["右向X"] = CosBJ(_____53C2_6570["方向角"] - 90)
    self["右向Y"] = SinBJ(_____53C2_6570["方向角"] - 90)
    self["创建提示圈"](self)
    self["Tick回调ID"] = addPeriodicCallback(_____53C2_6570["Tick间隔毫秒"] or 100, ____on_7EBF_6BB5_5371_9669_533ATick, self)
end
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype["推进"] = function(self, now)
    if self["已停止"] then
        return
    end
    if now >= self["到期Ms"] then
        self["停止"](self)
        return
    end
    local _____5355_4F4D_5217_8868 = self["参数"]["单位列表"](self["参数"]["变量"])
    do
        local i = 0
        while i < #_____5355_4F4D_5217_8868 do
            do
                local _____5355_4F4D = _____5355_4F4D_5217_8868[i + 1]
                if not isValidUnit(_____5355_4F4D) then
                    goto __continue9
                end
                self["推进单位"](self, now, _____5355_4F4D)
            end
            ::__continue9::
            i = i + 1
        end
    end
end
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    if self["Tick回调ID"] ~= 0 then
        removePeriodicCallback(self["Tick回调ID"])
        self["Tick回调ID"] = 0
    end
    if self["参数"]["on结束"] ~= nil then
        self["参数"]["on结束"]()
    end
end
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype["推进单位"] = function(self, now, _____5355_4F4D)
    local id = jass.GetHandleId(_____5355_4F4D)
    local dx = GetUnitX(_____5355_4F4D) - self["参数"]["起点X"]
    local dy = GetUnitY(_____5355_4F4D) - self["参数"]["起点Y"]
    local _____524D_5411_8DDD_79BB = dx * self["前向X"] + dy * self["前向Y"]
    local _____6A2A_5411_8DDD_79BB = dx * self["右向X"] + dy * self["右向Y"]
    local _____5F53_524D_5728_5185_90E8 = self["是否在内部"](
        self,
        GetUnitX(_____5355_4F4D),
        GetUnitY(_____5355_4F4D)
    )
    local _____72B6_6001 = self["单位状态表"][id]
    if _____72B6_6001 == nil then
        _____72B6_6001 = {
            ["在内部"] = false,
            ["下次周期Ms"] = 0,
            ["已采样"] = false,
            ["上次前向距离"] = _____524D_5411_8DDD_79BB,
            ["上次侧向符号"] = 0,
            ["下次穿越Ms"] = 0
        }
        self["单位状态表"][id] = _____72B6_6001
    end
    local _____5F53_524D_4FA7_5411_7B26_53F7 = _____6A2A_5411_8DDD_79BB < 0 and -1 or (_____6A2A_5411_8DDD_79BB > 0 and 1 or 0)
    local _____771F_6B63_8DE8_7EBF = _____72B6_6001["已采样"] and _____5F53_524D_4FA7_5411_7B26_53F7 ~= 0 and _____72B6_6001["上次侧向符号"] ~= 0 and _____5F53_524D_4FA7_5411_7B26_53F7 ~= _____72B6_6001["上次侧向符号"] and _____524D_5411_8DDD_79BB_5728_8303_56F4_5185(_____72B6_6001["上次前向距离"], self["参数"]["长度"]) and _____524D_5411_8DDD_79BB_5728_8303_56F4_5185(_____524D_5411_8DDD_79BB, self["参数"]["长度"]) and now >= _____72B6_6001["下次穿越Ms"]
    _____72B6_6001["已采样"] = true
    _____72B6_6001["上次前向距离"] = _____524D_5411_8DDD_79BB
    if _____5F53_524D_4FA7_5411_7B26_53F7 ~= 0 then
        _____72B6_6001["上次侧向符号"] = _____5F53_524D_4FA7_5411_7B26_53F7
    end
    if _____5F53_524D_5728_5185_90E8 and not _____72B6_6001["在内部"] then
        _____72B6_6001["在内部"] = true
        _____72B6_6001["下次周期Ms"] = now
        if self["参数"]["on进入"] ~= nil then
            self["参数"]["on进入"](_____5355_4F4D, self["参数"]["变量"])
        end
    elseif not _____5F53_524D_5728_5185_90E8 and _____72B6_6001["在内部"] then
        _____72B6_6001["在内部"] = false
        if self["参数"]["on离开"] ~= nil then
            self["参数"]["on离开"](_____5355_4F4D, self["参数"]["变量"])
        end
    end
    if _____771F_6B63_8DE8_7EBF then
        _____72B6_6001["下次穿越Ms"] = now + (self["参数"]["穿越防抖秒"] or 0) * 1000
        if self["参数"]["on穿越"] ~= nil then
            self["参数"]["on穿越"](_____5355_4F4D, self["参数"]["变量"])
        end
    end
    if _____5F53_524D_5728_5185_90E8 and self["参数"]["on周期"] ~= nil and now >= _____72B6_6001["下次周期Ms"] then
        self["参数"]["on周期"](_____5355_4F4D, self["参数"]["变量"])
        _____72B6_6001["下次周期Ms"] = now + (self["参数"]["周期秒"] or 1) * 1000
    end
end
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype["是否在内部"] = function(self, x, y)
    local dx = x - self["参数"]["起点X"]
    local dy = y - self["参数"]["起点Y"]
    local _____524D_5411_8DDD_79BB = dx * self["前向X"] + dy * self["前向Y"]
    if _____524D_5411_8DDD_79BB < 0 or _____524D_5411_8DDD_79BB > self["参数"]["长度"] then
        return false
    end
    local _____6A2A_5411_8DDD_79BB = dx * self["右向X"] + dy * self["右向Y"]
    return _____7EDD_5BF9_503C(_____6A2A_5411_8DDD_79BB) <= self["参数"]["宽度"] * 0.5
end
_____7EBF_6BB5_5371_9669_533A_5B9E_73B0.prototype["创建提示圈"] = function(self)
    if self["参数"]["提示圈"] == false then
        return
    end
    _____521B_5EFA_6280_80FD_63D0_793A_5708(__TS__ObjectAssign({
        ["类型"] = "矩形",
        ["宽度"] = self["参数"]["宽度"],
        ["长度"] = self["参数"]["长度"],
        ["朝向"] = self["参数"]["方向角"],
        ["持续时间"] = self["参数"]["持续秒"]
    }, self["参数"]["提示圈"] or ({}), {X = self["参数"]["起点X"], Y = self["参数"]["起点Y"]}))
end
____exports["创建线段危险区"] = function(_____53C2_6570)
    local _____5B9E_4F8B = __TS__New(_____7EBF_6BB5_5371_9669_533A_5B9E_73B0, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_3 = _____53C2_6570["清理"]
        ____self_3["登记清理"](
            ____self_3,
            _____53C2_6570["名称"],
            function()
                _____5B9E_4F8B["停止"](_____5B9E_4F8B)
            end
        )
    end
    return _____5B9E_4F8B
end
return ____exports
